AsyncDisplayKit/ASTextNode.mm

```diff
@@ -50,6 +50,77 @@
   UIColor *backgroundColor;
 };
 
+#pragma mark - ASTextKitRenderer
+
+// Not used at the moment but handy to have
+/*ASDISPLAYNODE_INLINE NSUInteger ASHashFromCGRect(CGRect rect)
+{
+  return ((*(NSUInteger *)&rect.origin.x << 10 ^ *(NSUInteger *)&rect.origin.y) + (*(NSUInteger *)&rect.size.width << 10 ^ *(NSUInteger *)&rect.size.height));
+}*/
+
+ASDISPLAYNODE_INLINE NSUInteger ASHashFromCGSize(CGSize size)
+{
+  return ((*(NSUInteger *)&size.width << 10 ^ *(NSUInteger *)&size.height));
+}
+
+@interface ASTextNodeRendererKey : NSObject
+@property (assign, nonatomic) ASTextKitAttributes attributes;
+@property (assign, nonatomic) CGSize constrainedSize;
+@end
+
+@implementation ASTextNodeRendererKey
+
+- (NSUInteger)hash
+{
+  return _attributes.hash() ^ ASHashFromCGSize(_constrainedSize);
+}
+
+- (BOOL)isEqual:(ASTextNodeRendererKey *)object
+{
+  if (self == object) {
+    return YES;
+  }
+  
+  return _attributes.hash() == object.attributes.hash()
+  && CGSizeEqualToSize(_constrainedSize, object.constrainedSize);
+}
+
+@end
+
+static NSCache *sharedRendererCache()
+{ 
+ static dispatch_once_t onceToken;
+ static NSCache *__rendererCache = nil;
+ dispatch_once(&onceToken, ^{
+   __rendererCache = [[NSCache alloc] init];
+   __rendererCache.countLimit = 500; // 500 renders cache
+ });
+ return __rendererCache;
+}
+
+/**
+ The concept here is that neither the node nor layout should ever have a strong reference to the renderer object.
+ This is to reduce memory load when loading thousands and thousands of text nodes into memory at once. Instead
+ we maintain a LRU renderer cache that is queried via stack-allocated keys.
```

Comment:
We have to remove the "stack-allocated keys" bit. 

I know you're already aware of this, but just to comment for any other readers, I think the cost of allocating these key-objects is acceptable for the time-being. In the future we can optimize this later if we find it to be a major source of contention, but this is the correct design and the implementation is simple & reliable.